docs/architecture/adr-074-timeout-params.md

```diff
@@ -0,0 +1,228 @@
+they have similar bandwidth and machine resources. The timings for propagating votes
+and blocks are likely to be reasonably similar across networks. As a result, the
+timeout parameters are good candidates for being hardcoded. Hardcoding the timeouts
+in Tendermint would mean entirely removing these parameters from any configuration
+that could be altered by either an application or a node operator. Instead,
+Tendermint would ship with a set of timeouts and all applications using Tendermint
+would use this exact same set of values.
+
+While Tendermint nodes often run with similar bandwidth and on similar cloud-hosted
+machines, there are enough points of variability to make configuring
+consensus timeouts meaningful. Namely, Tendermint network topologies are likely to be
+very different from chain to chain. Additionally, applications may vary greatly in 
+how long the `Commit` phase may take. Applications that perform more work during `Commit`
+require a longer `TimeoutCommit` to allow the application to complete its work
+and be prepared for the next height.
+
+* From https://github.com/tendermint/spec/issues/359 it looks like `TimeoutCommit` really only exists for the purpose of allowing
+precommits to be collected after the +2/3 has initially been seen. Not much more value to it than that. 
+
+* allowing long timeout commit discussed here: https://github.com/tendermint/tendermint/issues/5911#issuecomment-973560381
+
+## Decision
+
+None
+
+## Detailed Design
+
+### New Consensus Parameters
+
+A new `TimeoutParams` `message` will be added to the [params.proto file][consensus-params-proto].
+This message will have the following form:
+
+```proto
+message TimeoutParams {
+ google.protobuf.Duration timeout_propose = 1
+       [(gogoproto.nullable) = false, (gogoproto.stdduration) = true];
+ google.protobuf.Duration timeout_propose_delta = 2
+       [(gogoproto.nullable) = false, (gogoproto.stdduration) = true];
+ google.protobuf.Duration timeout_vote = 3
+       [(gogoproto.nullable) = false, (gogoproto.stdduration) = true];
+ google.protobuf.Duration timeout_vote_delta = 4
+       [(gogoproto.nullable) = false, (gogoproto.stdduration) = true];
+ google.protobuf.Duration timeout_commit = 5
+       [(gogoproto.nullable) = false, (gogoproto.stdduration) = true];
+ bool skip_timeout_commit = 6;
+}
+```
+
+This new `message` will then be added as a field into the [`ConsensusParams`
+message][consensus-params-proto]. The same default values that are [currently
+set for these parameters][current-timeout-defaults] in the local configuration
+file will be used as the defaults for these new consensus parameters in the
+[consensus parameter defaults][default-consensus-params].
+
+Validation of these new parameters will be identical to their [validation
+currently][time-param-validation].
+Namely, these parameters must be non-negative.
+
+### Removal of Old Parameters
+
+The old `timeout-*` parameters that are configured in the [config.toml][config-toml]
```

Comment:
We should probably stage this somehow, even if we don't allow temporary overrides. For example:

1. In release v0.36, a node receiving a block without timeouts in the consensus parameters will fall back on the local settings (this allows node software to be upgraded without stopping the chain)
2. In release v0.36, if any non-default timeout values are set in the config, the node will log a warning when it starts up about the upcoming deprecation of those values.
3. In release v0.37, the parameters are suppressed or removed, so that starting up a node which has them in its config will print a specific error pointing to the documentation (rather than just "unknown field lol")

(This is just a sketch of how it could be done)

The other thing I wonder is how do we bootstrap these values into consensus parameters, on chains that don't already have them? For the defaults, of course, it's a noop. But if I wanted to have a longer block delay, say, can we do that as a "soft" migration?

Reply:
> The other thing I wonder is how do we bootstrap these values into consensus parameters, on chains that don't already have them? For the defaults, of course, it's a noop. But if I wanted to have a longer block delay, say, can we do that as a "soft" migration?

Unless I'm misunderstanding the question, this should be pretty straightforward. The values themselves are not actually included in the blocks, just a hash of the values. The hash already exists, so we just have to make sure to make the change in a way that preserves the old hashes.